pkg/phlaredb/compact.go

```diff
@@ -0,0 +1,301 @@
+package phlaredb
+
+import (
+	"context"
+	"math"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/segmentio/parquet-go"
+
+	"github.com/grafana/phlare/pkg/iter"
+	phlaremodel "github.com/grafana/phlare/pkg/model"
+	phlareparquet "github.com/grafana/phlare/pkg/parquet"
+	"github.com/grafana/phlare/pkg/phlaredb/block"
+	schemav1 "github.com/grafana/phlare/pkg/phlaredb/schemas/v1"
+	"github.com/grafana/phlare/pkg/phlaredb/tsdb/index"
+	"github.com/grafana/phlare/pkg/util"
+	"github.com/grafana/phlare/pkg/util/loser"
+)
+
+type BlockReader interface {
+	Profiles() []parquet.RowGroup
+	Index() IndexReader
+	// Symbols() SymbolReader
+}
+
+type SymbolReader interface {
+	// todo
+}
+
+func Compact(ctx context.Context, src []BlockReader, dst string) (block.Meta, error) {
+	meta := block.NewMeta()
+	blockPath := filepath.Join(dst, meta.ULID.String())
+	if err := os.MkdirAll(blockPath, 0o777); err != nil {
+		return block.Meta{}, err
+	}
+	indexPath := filepath.Join(blockPath, block.IndexFilename)
+	indexw, err := prepareIndexWriter(ctx, indexPath, src)
+	if err != nil {
+		return block.Meta{}, err
+	}
+	profilePath := filepath.Join(blockPath, (&schemav1.ProfilePersister{}).Name()+block.ParquetSuffix)
+	profileFile, err := os.OpenFile(profilePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o644)
+	if err != nil {
+		return block.Meta{}, err
+	}
+	profileWriter := newProfileWriter(profileFile)
+
+	// todo new symbdb
```

Comment:
@kolesnikovae see how we wrap the rowIterator to remap each rows value...

I'm leaving symdb alone for now as you're working on this, I'll start testing and benchmarking this for now.